### PR TITLE
SoundWire: introduce hw_ops and move auxdevice handling to dedicated file

### PR DESCRIPTION
**Pull request by plbossart**
This PR adds yet another set of cleanups
a) all the helpers are now exposed as hw_ops callbacks
b) the auxiliary device code is now completely generic.

This will allow for easier introduction of new hw_ops implementations in the future without throwing away all the goodness of the auxiliary device handling. Since there will be a tighter coupling with HDaudio, it's likely that future hw_ops are implemented in the SOF driver directly.

### CODE REVIEW

**Review by bardliao — DISMISSED** [review]

File: drivers/soundwire/intel.c [hunk 1]
```diff
@@ -745,10 +745,10 @@ static int intel_free_stream(struct sdw_intel *sdw,
  * bank switch routines
  */
 
-static int intel_pre_bank_switch(struct sdw_bus *bus)
```
Comment (ranj063) [inline drivers/soundwire/intel.c:748]: @plbossart i  know i may be asking for too much but if you had a separate patch for using each new op, that would make it easier to tell if something was done wrong easily

Comment (plbossart) [reply drivers/soundwire/intel.c:748]: humm, why not. It's true that the next-to-last patch is quite invasive. That would add about 10 patches but they'd be smaller.

Comment (plbossart) [reply drivers/soundwire/intel.c:748]: done, this is simpler indeed to review, thanks for the suggestion @ranj063

**Comment by plbossart** [comment]
I'll rebase when PR https://github.com/thesofproject/linux/pull/3933 is merged, and remove the draft status.

File: drivers/soundwire/intel.h [hunk 2]
```diff
@@ -47,15 +50,100 @@ struct sdw_intel {
...
+static inline void sdw_intel_debugfs_exit(struct sdw_intel *sdw)
+{
+	if (sdw && sdw->link_res && sdw->link_res->hw_ops &&
+	    sdw->link_res->hw_ops->debugfs_exit)
+		sdw->link_res->hw_ops->debugfs_exit(sdw);
+}
+
+static inline int sdw_intel_register_dai(struct sdw_intel *sdw)
+{
+	if (sdw && sdw->link_res &&  sdw->link_res->hw_ops &&
```
Comment (RanderWang) [inline drivers/soundwire/intel.h]: one more whitespace before sdw->link_res->hw_ops.

File: drivers/soundwire/intel.h [hunk 3]
```diff
@@ -47,15 +50,100 @@ struct sdw_intel {
...
+static inline void sdw_intel_debugfs_init(struct sdw_intel *sdw)
+{
+	if (sdw && sdw->link_res && sdw->link_res->hw_ops &&
+	    sdw->link_res->hw_ops->debugfs_init)
+		sdw->link_res->hw_ops->debugfs_init(sdw);
+}
+
+static inline void sdw_intel_debugfs_exit(struct sdw_intel *sdw)
+{
+	if (sdw && sdw->link_res && sdw->link_res->hw_ops &&
```
Comment (RanderWang) [inline drivers/soundwire/intel.h]: the count of "sdw && sdw->link_res && sdw->link_res->hw_ops" is 11, can we define a macro ?

File: drivers/soundwire/intel_auxdevice.c [hunk 4]
```diff
@@ -0,0 +1,679 @@
...
+
+static int md_flags;
+module_param_named(sdw_md_flags, md_flags, int, 0444);
+MODULE_PARM_DESC(sdw_md_flags, "SoundWire Intel Master device flags (0x0 all off)");
+
+static int generic_pre_bank_switch(struct sdw_bus *bus)
+{
+	struct sdw_cdns *cdns = bus_to_cdns(bus);
+	struct sdw_intel *sdw = cdns_to_intel(cdns);
+
```
Comment (RanderWang) [inline drivers/soundwire/intel_auxdevice.c:51]: don't need to check "sdw && sdw->link_res && sdw->link_res->hw_ops" ?

File: drivers/soundwire/intel.h (hunk 2 above)
Comment (bardliao) [reply drivers/soundwire/intel.h]: > one more whitespace before sdw->link_res->hw_ops.

In fact, the double whitespace is seen in every `if (sdw && sdw->link_res &&  sdw->link_res->hw_ops &&`. ;)

File: drivers/soundwire/intel.h (hunk 3 above)
Comment (plbossart) [reply drivers/soundwire/intel.h]: done, thanks for the suggestion @RanderWang

File: drivers/soundwire/intel.h (hunk 2 above)
Comment (plbossart) [reply drivers/soundwire/intel.h]: fixed with macros

**Comment by plbossart** [comment]
v2 - comments from @RanderWang addressed 

````diff
diff --git a/drivers/soundwire/intel.h b/drivers/soundwire/intel.h
index 7eb48e459f6b..de9883313c8f 100644
--- a/drivers/soundwire/intel.h
+++ b/drivers/soundwire/intel.h
@@ -54,96 +54,88 @@ struct sdw_intel {
 
 #define INTEL_MASTER_RESET_ITERATIONS	10
 
+#define SDW_INTEL_CHECK_OPS(sdw, cb)	((sdw) && (sdw)->link_res && (sdw)->link_res->hw_ops && \
+					 (sdw)->link_res->hw_ops->cb)
+#define SDW_INTEL_OPS(sdw, cb)		((sdw)->link_res->hw_ops->cb)
+
 static inline void sdw_intel_debugfs_init(struct sdw_intel *sdw)
 {
-	if (sdw && sdw->link_res && sdw->link_res->hw_ops &&
-	    sdw->link_res->hw_ops->debugfs_init)
-		sdw->link_res->hw_ops->debugfs_init(sdw);
+	if (SDW_INTEL_CHECK_OPS(sdw, debugfs_init))
+		SDW_INTEL_OPS(sdw, debugfs_init)(sdw);
 }
 
 static inline void sdw_intel_debugfs_exit(struct sdw_intel *sdw)
 {
-	if (sdw && sdw->link_res && sdw->link_res->hw_ops &&
-	    sdw->link_res->hw_ops->debugfs_exit)
-		sdw->link_res->hw_ops->debugfs_exit(sdw);
+	if (SDW_INTEL_CHECK_OPS(sdw, debugfs_exit))
+		SDW_INTEL_OPS(sdw, debugfs_exit)(sdw);
 }
 
 static inline int sdw_intel_register_dai(struct sdw_intel *sdw)
 {
-	if (sdw && sdw->link_res &&  sdw->link_res->hw_ops &&
-	    sdw->link_res->hw_ops->register_dai)
-		return sdw->link_res->hw_ops->register_dai(sdw);
+	if (SDW_INTEL_CHECK_OPS(sdw, register_dai))
+		return SDW_INTEL_OPS(sdw, register_dai)(sdw);
 	return -ENOTSUPP;
 }
 
 static inline void sdw_intel_check_clock_stop(struct sdw_intel *sdw)
 {
-	if (sdw && sdw->link_res &&  sdw->link_res->hw_ops &&
-	    sdw->link_res->hw_ops->check_clock_stop)
-		sdw->link_res->hw_ops->check_clock_stop(sdw);
+	if (SDW_INTEL_CHECK_OPS(sdw, check_clock_stop))
+		SDW_INTEL_OPS(sdw, check_clock_stop)(sdw);
 }
 
 static inline int sdw_intel_start_bus(struct sdw_intel *sdw)
 {
-	if (sdw && sdw->link_res &&  sdw->link_res->hw_ops &&
-	    sdw->link_res->hw_ops->start_bus)
-		return sdw->link_res->hw_ops->start_bus(sdw);
+	if (SDW_INTEL_CHECK_OPS(sdw, start_bus))
+		return SDW_INTEL_OPS(sdw, start_bus)(sdw);
 	return -ENOTSUPP;
 }
 
 static inline int sdw_intel_start_bus_after_reset(struct sdw_intel *sdw)
 {
-	if (sdw && sdw->link_res &&  sdw->link_res->hw_ops &&
-	    sdw->link_res->hw_ops->start_bus_after_reset)
-		return sdw->link_res->hw_ops->start_bus_after_reset(sdw);
+	if (SDW_INTEL_CHECK_OPS(sdw, start_bus_after_reset))
+		return SDW_INTEL_OPS(sdw, start_bus_after_reset)(sdw);
 	return -ENOTSUPP;
 }
 
 static inline int sdw_intel_start_bus_after_clock_stop(struct sdw_intel *sdw)
 {
-	if (sdw && sdw->link_res &&  sdw->link_res->hw_ops &&
-	    sdw->link_res->hw_ops->start_bus_after_clock_stop)
-		return sdw->link_res->hw_ops->start_bus_after_clock_stop(sdw);
+	if (SDW_INTEL_CHECK_OPS(sdw, start_bus_after_clock_stop))
+		return SDW_INTEL_OPS(sdw, start_bus_after_clock_stop)(sdw);
 	return -ENOTSUPP;
 }
 
 static inline int sdw_intel_stop_bus(struct sdw_intel *sdw, bool clock_stop)
 {
-	if (sdw && sdw->link_res &&  sdw->link_res->hw_ops &&
-	    sdw->link_res->hw_ops->stop_bus)
-		return sdw->link_res->hw_ops->stop_bus(sdw, clock_stop);
+	if (SDW_INTEL_CHECK_OPS(sdw, stop_bus))
+		return SDW_INTEL_OPS(sdw, stop_bus)(sdw, clock_stop);
 	return -ENOTSUPP;
 }
 
 static inline int sdw_intel_link_power_up(struct sdw_intel *sdw)
 {
-	if (sdw && sdw->link_res && sdw->link_res->hw_ops &&
-	    sdw->link_res->hw_ops->link_power_up)
-		return sdw->link_res->hw_ops->link_power_up(sdw);
+	if (SDW_INTEL_CHECK_OPS(sdw, link_power_up))
+		return SDW_INTEL_OPS(sdw, link_power_up)(sdw);
 	return -ENOTSUPP;
 }
 
 static inline int sdw_intel_link_power_down(struct sdw_intel *sdw)
 {
-	if (sdw && sdw->link_res &&  sdw->link_res->hw_ops &&
-	    sdw->link_res->hw_ops->link_power_down)
-		return sdw->link_res->hw_ops->link_power_down(sdw);
+	if (SDW_INTEL_CHECK_OPS(sdw, link_power_down))
+		return SDW_INTEL_OPS(sdw, link_power_down)(sdw);
 	return -ENOTSUPP;
 }
 
 static inline int sdw_intel_shim_check_wake(struct sdw_intel *sdw)
 {
-	if (sdw && sdw->link_res &&  sdw->link_res->hw_ops &&
-	    sdw->link_res->hw_ops->shim_check_wake)
-		return sdw->link_res->hw_ops->shim_check_wake(sdw);
+	if (SDW_INTEL_CHECK_OPS(sdw, shim_check_wake))
+		return SDW_INTEL_OPS(sdw, shim_check_wake)(sdw);
 	return -ENOTSUPP;
 }
 
 static inline void sdw_intel_shim_wake(struct sdw_intel *sdw, bool wake_enable)
 {
-	if (sdw && sdw->link_res &&  sdw->link_res->hw_ops &&
-	    sdw->link_res->hw_ops->shim_wake)
-		sdw->link_res->hw_ops->shim_wake(sdw, wake_enable);
+	if (SDW_INTEL_CHECK_OPS(sdw, shim_wake))
+		SDW_INTEL_OPS(sdw, shim_wake)(sdw, wake_enable);
 }
 
 #endif /* __SDW_INTEL_LOCAL_H */
diff --git a/drivers/soundwire/intel_auxdevice.c b/drivers/soundwire/intel_auxdevice.c
index 5f192b0cc79e..96c6b2112feb 100644
--- a/drivers/soundwire/intel_auxdevice.c
+++ b/drivers/soundwire/intel_auxdevice.c
@@ -613,7 +613,6 @@ static int __maybe_unused intel_resume_runtime(struct device *dev)
 			return ret;
 		}
 
-
 	} else if (clock_stop_quirks & SDW_INTEL_CLK_STOP_BUS_RESET) {
 		ret = sdw_intel_link_power_up(sdw);
 		if (ret) {

````

**Review by bardliao — APPROVED** [review]

**Review by RanderWang — APPROVED** [review]
LGTM

**Pull request merged**